File: large_file_finder.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"sync"
)

const (
	sizeLimitMB = 50
	sizeLimitBytes = sizeLimitMB * 1024 * 1024
)

type FileInfo struct {
	Path string
	Size int64
}

func parseGitignore(gitignorePath string) ([]string, error) {
	file, err := os.Open(gitignorePath)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	var patterns []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line != "" && !strings.HasPrefix(line, "#") {
			patterns = append(patterns, line)
		}
	}
	return patterns, scanner.Err()
}

func isIgnored(path string, ignorePatterns []string) bool {
	for _, pattern := range ignorePatterns {
		matched, err := filepath.Match(pattern, filepath.Base(path))
		if err == nil && matched {
			return true
		}
		matched, err = filepath.Match(pattern, path)
		if err == nil && matched {
			return true
		}
	}
	return false
}

func findLargeFiles(dir string, ignorePatterns []string) []FileInfo {
	var largeFiles []FileInfo
	var mutex sync.Mutex
	var wg sync.WaitGroup

	filepath.Walk(dir, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}

		if info.IsDir() {
			if isIgnored(path, ignorePatterns) {
				return filepath.SkipDir
			}
			return nil
		}

		if !isIgnored(path, ignorePatterns) && info.Size() > sizeLimitBytes {
			wg.Add(1)
			go func() {
				defer wg.Done()
				mutex.Lock()
				largeFiles = append(largeFiles, FileInfo{Path: path, Size: info.Size()})
				mutex.Unlock()
			}()
		}

		return nil
	})

	wg.Wait()
	return largeFiles
}

func main() {
	var dir string
	if len(os.Args) > 1 {
		dir = os.Args[1]
	} else {
		var err error
		dir, err = os.Getwd()
		if err != nil {
			fmt.Println("Error getting current directory:", err)
			return
		}
	}

	gitignorePath := filepath.Join(dir, ".gitignore")
	ignorePatterns, err := parseGitignore(gitignorePath)
	if err != nil {
		fmt.Println("Warning: Could not parse .gitignore:", err)
	}

	fmt.Printf("Searching for files larger than %d MiB in: %s\n", sizeLimitMB, dir)
	fmt.Println("(Respecting .gitignore rules)")

	largeFiles := findLargeFiles(dir, ignorePatterns)

	if len(largeFiles) > 0 {
		fmt.Println("\nLarge files found:")
		sort.Slice(largeFiles, func(i, j int) bool {
			return largeFiles[i].Size > largeFiles[j].Size
		})
		for _, file := range largeFiles {
			fmt.Printf("%s: %.2f MiB\n", file.Path, float64(file.Size)/(1024*1024))
		}
	} else {
		fmt.Printf("\nNo files larger than %d MiB found.\n", sizeLimitMB)
	}

	fmt.Printf("\nTotal number of large files: %d\n", len(largeFiles))
}

File: large_file_finder.py
import os
import sys
from pathlib import Path
import fnmatch
from typing import List, Tuple


def parse_gitignore(gitignore_path: Path) -> List[str]:
    if not gitignore_path.exists():
        return []

    with open(gitignore_path, "r") as f:
        return [line.strip() for line in f if line.strip() and not line.startswith("#")]


def is_ignored(path: Path, ignore_patterns: List[str]) -> bool:
    for pattern in ignore_patterns:
        if fnmatch.fnmatch(str(path), pattern) or fnmatch.fnmatch(path.name, pattern):
            return True
    return False


def find_large_files(
    directory: Path, size_limit_mb: int = 50, ignore_patterns: List[str] = []
) -> List[Tuple[Path, int]]:
    size_limit_bytes = size_limit_mb * 1024 * 1024  # Convert MB to bytes
    large_files = []

    for root, dirs, files in os.walk(directory):
        # Remove ignored directories
        dirs[:] = [d for d in dirs if not is_ignored(Path(root) / d, ignore_patterns)]

        for file in files:
            file_path = Path(root) / file
            if file_path.is_file() and not is_ignored(file_path, ignore_patterns):
                size = file_path.stat().st_size
                if size > size_limit_bytes:
                    large_files.append((file_path, size))

    return large_files


def main():
    if len(sys.argv) > 1:
        directory = Path(sys.argv[1])
    else:
        directory = Path.cwd()

    gitignore_path = directory / ".gitignore"
    ignore_patterns = parse_gitignore(gitignore_path)

    print(f"Searching for files larger than 50 MiB in: {directory}")
    print("(Respecting .gitignore rules)")
    large_files = find_large_files(directory, ignore_patterns=ignore_patterns)

    if large_files:
        print("\nLarge files found:")
        for file_path, size in sorted(large_files, key=lambda x: x[1], reverse=True):
            print(f"{file_path}: {size / (1024 * 1024):.2f} MiB")
    else:
        print("\nNo files larger than 50 MiB found.")

    print(f"\nTotal number of large files: {len(large_files)}")


if __name__ == "__main__":
    main()
